src/libcore/iter.rs

```diff
@@ -1312,10 +1320,15 @@ pub trait Iterator {
         let mut other = other.into_iter();
 
         loop {
-            match (self.next(), other.next()) {
-                (None, None) => return true,
-                (None, _) | (_, None) => return false,
-                (Some(x), Some(y)) => if x != y { return false },
+            let (l, r) = (self.next(), other.next());
+            return if l.is_none() && r.is_none() {
+                true
+            } else if l.is_none() || r.is_none() {
+                false
+            } else if PartialEq::ne(&l.unwrap(), &r.unwrap()) {
```

Comment:
`if let` desugars into a `match`, so doing something like `else if let (Some(x), Some(y))` would basically return us to square one.